File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/SetLocaleMiddleware.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Globalization;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;

namespace Microsoft.Bot.Solutions.Middleware
{
    /// <summary>
    /// Set locale by user input locale.
    /// </summary>
    public class SetLocaleMiddleware : IMiddleware
    {
        private readonly string defaultLocale;

        public SetLocaleMiddleware(string defaultDefaultLocale)
        {
            this.defaultLocale = defaultDefaultLocale;
        }

        public async Task OnTurnAsync(ITurnContext context, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
        {
            var cultureInfo = !string.IsNullOrWhiteSpace(context.Activity.Locale) ? new CultureInfo(context.Activity.Locale) : new CultureInfo(this.defaultLocale);

            CultureInfo.CurrentUICulture = CultureInfo.CurrentCulture = cultureInfo;

            await next(cancellationToken).ConfigureAwait(false);
        }
    }
}

File: sdk/typescript/libraries/bot-solutions/src/middleware/setLocaleMiddleware.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { Middleware, TurnContext } from 'botbuilder';

/**
 * Set locale by user input locale.
 */
export class SetLocaleMiddleware implements Middleware {
    private readonly defaultLocale: string;

    public constructor(defaultLocale: string) {
        if (defaultLocale === undefined) { throw new Error (`Parameter 'defaultLocale' cannot be undefined.`); }
        this.defaultLocale = defaultLocale;
    }

    public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
        const cultureInfo: string = context.activity.locale || this.defaultLocale;
        context.activity.locale = cultureInfo.toLowerCase();

        await next();
    }
}
